### PR TITLE
add StrictTree

### PR DESCRIPTION
**Pull request by shawjef3**
This is a strict version of Tree. Maybe some people would find it useful.

It's largely a copy of Tree, which I altered for maximum strictness.

### CODE REVIEW

**Comment by shawjef3** [comment]
Something that needs improving is the flatten method. The naive way, which is something like rootlabel +: subforest.flatMap(_.flatten), causes a stack overflow.

**Comment by jbgi** [comment]
Would additional strict data constructors for Tree be a lighter solution?

File: core/src/main/scala/scalaz/StrictTree.scala [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    def alignWith[A, B, C](f: (\&/[A, B]) ⇒ C): (StrictTree[A], StrictTree[B]) => StrictTree[C] = {
+      def align(ta: StrictTree[A], tb: StrictTree[B]): StrictTree[C] =
+        StrictTree.Node(f(\&/(ta.rootLabel, tb.rootLabel)), Align[Vector].alignWith[StrictTree[A], StrictTree[B], StrictTree[C]]({
+          case \&/.This(sta) ⇒ sta map {a ⇒ f(\&/.This(a))}
+          case \&/.That(stb) ⇒ stb map {b ⇒ f(\&/.That(b))}
+          case \&/(sta, stb) ⇒ align(sta, stb)
+        })(ta.subForest, tb.subForest))
+      align _
+    }
+    def zip[A, B](a: StrictTree[A], b: StrictTree[B]): StrictTree[(A, B)] = {
```
Comment (xuwei-k) [inline core/src/main/scala/scalaz/StrictTree.scala]: unused method?

**Comment by shawjef3** [comment]
@jbgi I don't think strict constructors for Tree would be enough. Tree makes heavy use of lazy val, call by name, and Stream that would make it somewhat lazy even if the constructor wasn't..

File: tests/jvm/src/test/scala/scalaz/StrictTreeTestJVM.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package scalaz
+
+import scalaz.scalacheck.ScalazArbitrary._
+import StrictTree._
+import org.scalacheck.Gen
+import org.scalacheck.Prop.forAll
+
+object StrictTreeTestJVM extends SpecLite {
+
+  "ScalazArbitrary.treeGenSized" ! forAll(Gen.choose(1, 200)){ size =>
```
Comment (xuwei-k) [inline tests/jvm/src/test/scala/scalaz/StrictTreeTestJVM.scala]: s/treeGenSized/strictTreeGenSized

File: core/src/main/scala/scalaz/StrictTree.scala [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        A.order(x.rootLabel, y.rootLabel) match {
+          case Ordering.EQ =>
+            Order[Vector[StrictTree[A]]].order(x.subForest, y.subForest)
+          case x => x
+        }
+    }
+
+  /* TODO
+  def applic[A, B](f: StrictTree[A => B]) = a => StrictTree.node((f.rootLabel)(a.rootLabel), implicitly[Applic[newtypes.ZipVector]].applic(f.subForest.map(applic[A, B](_)).?)(a.subForest ?).value)
+   */
```
Comment (aloiscochard) [inline core/src/main/scala/scalaz/StrictTree.scala]: @shawjef3 do you plan to implement it as part of this PR, or maybe just stale dead code?

Comment (shawjef3) [inline core/src/main/scala/scalaz/StrictTree.scala]: @aloiscochard It's a TODO for Tree, so I copied it to StrictTree. You'd have to ask @retronym what the status is.

**Comment by aloiscochard** [comment]
@jbgi and even if we get the same semantic with strict constructor, the performance would be probably much different simply because of the lazy val (run-time check at read).

@shawjef3 Great stuff, thanks for the contribution!

just curious what was the original motivation, was it performance?

**Comment by shawjef3** [comment]
@aloiscochard Yes, basically the motivation is performance.

**Comment by shawjef3** [comment]
Something I'm leaving for a later PR is Tree#toStrictTree. I'm having a heck of a time implementing it without causing a stack overflow. Something about how I'm using Trampoline is incorrect.

**Comment by shawjef3** [comment]
I added Tree#toStrictTree, and made sure that it works with deeply nested trees.

StrictTree#toTree had a performance problem with deeply nested trees due to a huge number of calls to Vector#flatten, so I changed it to an imperative implementation. Likewise for StrictTree#flatten.

I found that the Equals instance for Tree does not work for deeply nested trees, which causes some problems with some tests. I changed it to a trampolined implementation. Likewise for StrictTree's Equals.

**Comment by shawjef3** [comment]
A lot of methods in StrictTree have problems with stack overflows for deep trees. I'm working on new implementations, but they can wait for a later release.

The only other thing I've noticed is that my implementation of loc was premature. It needs to be removed. People can still use StrictTree#toTree#loc. StrictTree#loc should be implemented later to return a strict version of a TreeLoc. I'll upload a new change in a few minutes

**Comment by shawjef3** [comment]
I replaced several trampolined implementations with procedural implementations, yielding large speed increases.

Is there a way to squash this pull request without a force push?

**Comment by shawjef3** [comment]
I think I'm done with this for now. traverse1 and cobind still can cause stack overflow exceptions, but they're going to be harder to convert to stackless implementations.

**Comment by xuwei-k** [comment]
> Is there a way to squash this pull request without a force push?

There is no way. Please rebase and squash commits.

**Comment by xuwei-k** [comment]
I think `BottomUpStackElem ` and `ZipStackElem` should define in companion object, not `class StrictTree`.

File: core/src/main/scala/scalaz/StrictTree.scala [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+  /**
+    * This implementation is 24x faster than the trampolined implementation for StrictTreeTestJVM's hashCode test.
+    *
+    * @return
+    */
+  override def hashCode(): Int = {
+    runBottomUp(hashCodeCombiner)
+  }
+
+  override def equals(obj: scala.Any): Boolean = {
```
Comment (xuwei-k) [inline core/src/main/scala/scalaz/StrictTree.scala]: unnecessary override?

Comment (shawjef3) [inline core/src/main/scala/scalaz/StrictTree.scala]: The reason for this override is to avoid stack overflow on recursive calls to equals.

**Comment by shawjef3** [comment]
@xuwei-k I moved the classes as you requested.

**Comment by shawjef3** [comment]
I also moved AlignStackElem,

**Comment by shawjef3** [comment]
What is keeping this from being accepted?

**Comment by aloiscochard** [comment]
:+1:

**Comment by shawjef3** [comment]
@aloiscochard @xuwei-k mind if I merge https://github.com/shawjef3/scalaz/commit/e9c833911dadffee5b0067adfb60a5ff448add66 into this pull request? Its purpose is to make StrictTree.unfoldTree and unfoldForest more strict. StrictTree doesn't benefit from the children being created by continuations.

**Comment by aloiscochard** [comment]
@shawjef3 makes sense :+1:

**Comment by TomasMikula** [comment]
Has anyone else experienced tests taking too long or even `OutOfMemoryError` during tests after this was merged?

**Comment by vmarquez** [comment]
val size = 1000000



Might be a little overkill.

**Pull request merged**